Ironic overcloud hypervisor deployment (#1089)
stackhpc/stackhpc-kayobe-config: doc/source/configuration/ironic.rst

Reviewer: Generally I think the page should be a bit more declarative, at the moment it's written in a future tense. For example, I think that:

> Overcloud Ironic **will be** deployed with a listening TFTP server on the control plane

Would flow better as

> Overcloud Ironic **is** deployed with a listening TFTP server on the control plane

I don't think it would be a lot of effort to update (basically a find & replace: `will be -> is`) but it's a bit of a nitpick so if you want to leave it as is then that's fine

    allocation_pool_end: "10.10.3.3"

  openstack_routers:
    - "{{ openstack_router_ironic }}"

  openstack_router_ironic:
    - name: ironic
      project: admin
      interfaces:
        - net: "provision-net"
          subnet: "provision-net"
          portip: "172.0.1.1"
        - net: "cleaning-net"
          subnet: "cleaning-net"
          portip: "172.1.0.1"
          network: internal-net

To provision baremetal nodes in Nova you will also require setting a flavour
specific to that type of baremetal host. You will need to replace the custom
resource ``resources:CUSTOM_<YOUR_BAREMETAL_RESOURCE_CLASS>`` placeholder with
the resource class of your baremetal hosts, you will also need this later when
configuring the baremetal-compute inventory.

.. code-block:: yaml

  openstack_flavors:
    - "{{ openstack_flavor_baremetal_A }}"
    # Bare metal compute node.
    openstack_flavor_baremetal_A:
    name: "baremetal-A"
    ram: 1048576
    disk: 480
    vcpus: 256
    extra_specs:
    "resources:CUSTOM_<YOUR_BAREMETAL_RESOURCE_CLASS>": 1
    "resources:VCPU": 0
    "resources:MEMORY_MB": 0
    "resources:DISK_GB": 0

Enabling conntrack
==================

> UEFI booting requires conntrack_helper to be configured on the Ironic neutron

Reviewer 2: Only when using ML2/OVS

Reviewer 3: IIRC also only using firewall_driver = iptables_hybrid,, NOT firewall_driver = openvswitch

Reviewer: Done

> router, this is due to TFTP traffic being dropped due to being UDP. You will

Reviewer 4: ```suggestion
router, otherwise TFTP traffic is dropped due to it being UDP. You will
```
I had to re-read this a couple of times to understand it. Would suggest simplifying the statement a bit :)

conntrack_helper.

.. code-block:: json

  "conntrack_helpers": {
    "protocol": "udp",
    "port": 69,
    "helper": "tftp"
  }

Currently it's not possible to add this helper via the OpenStack CLI, to add
this to the Ironic router you will need to make a request to the API directly,
for example via cURL.

.. code-block:: console

  curl -g -i -X POST \
  http://<internal_api_vip>:9696/v2.0/routers/<ironic_router_uuid>/conntrack_helpers \
  -H "Accept: application/json" \
  -H "User-Agent: openstacksdk/2.0.0 keystoneauth1/5.4.0 python-requests/2.31.0 CPython/3.9.18" \
  -H "X-Auth-Token: <issued_token>" \
  -d '{ "conntrack_helper": {"helper": "tftp", "protocol": "udp", "port": 69 } }'

TFTP server
===========

By default the Ironic TFTP server (ironic_pxe container) will call the UEFI
boot file ``ipxe-x86_64.efi`` instead of ``ipxe.efi`` meaning no boot file will
be sent during the PXE boot process in the default configuration.

As of now this is solved by using a hack workaround by changing the boot file
in the ``ironic_pxe`` container. To do this you will need to enter the
container and rename the file manually.

.. code-block:: console

>   docker exec ironic_pxe “mv /tftpboot/ipxe-x86_64.efi /tftpboot/ipxe.efi”

Reviewer 4: More a general question then a review comment, but are we looking at any ways to automate this? It sounds like we'd be relying on these to be done consistently. Some hooks in Kayobe, if not patches to Ironic upstream.

Reviewer 5: The conntrack CLI workaround shouldn't be necessary anymore, so will need to be updated. Most of this documentation was written from working on a Train client, I don't think this workaround has been necessary since [Xena](https://github.com/openstack/python-openstackclient/commit/fa8c8d26a7696d169b0b9d5aaf6b723d8feee08a) and we can use a dedicated CLI command for this now.

I think the TFTP workaround was fixed in this [commit](https://github.com/openstack/kolla/commit/1bb57ae116b4b8931bfcd512038b436f77c67e0c) which will create a symlink for ipxe.efi to whatever architecture is used, and is in Yoga and later.

Reviewer 4: Makes sense to have dropped these sections then, given we're targetting 2023.1 here.
Perhaps we can still save these instructions on a wiki page, just in case another pre-Yoga deployment comes along?

======
Ironic
======

Experimental documentation
==========================

> This documentation is still WIP and is subject to change.

Reviewer: Sphinx has support for real [warning banners](https://sublime-and-sphinx-guide.readthedocs.io/en/latest/notes_warnings.html#warnings), this would be a good spot to use one:

```suggestion
.. warning::
  This page is a work in progress, and is subject to change.
```

        - net: "cleaning-net"
          subnet: "cleaning-net"
          portip: "172.1.0.1"
          network: internal-net

To provision baremetal nodes in Nova you will also require setting a flavour
specific to that type of baremetal host. You will need to replace the custom
resource ``resources:CUSTOM_<YOUR_BAREMETAL_RESOURCE_CLASS>`` placeholder with
the resource class of your baremetal hosts, you will also need this later when
configuring the baremetal-compute inventory.

.. code-block:: yaml

  openstack_flavors:
    - "{{ openstack_flavor_baremetal_A }}"
    # Bare metal compute node.
    openstack_flavor_baremetal_A:
    name: "baremetal-A"
    ram: 1048576
    disk: 480
    vcpus: 256
    extra_specs:
      "resources:CUSTOM_<YOUR_BAREMETAL_RESOURCE_CLASS>": 1
      "resources:VCPU": 0
      "resources:MEMORY_MB": 0
      "resources:DISK_GB": 0

Enabling conntrack (ML2/OVS only)
=================================

Conntrack_helper will be required when UEFI booting on a cloud with ML2/OVS
and using the iptables firewall_driver, otherwise TFTP traffic is dropped due
to it being UDP. You will need to define some extension drivers in ``neutron.yml``

> to ensure conntrack is enabled in neutron server.

Reviewer: You've done well to capitalise `Nova` and `Ironic` but missed a few instances of `OpenStack`, `Neutron`, and `Kayobe`
This is just one instance, but rather than pointing every one of them out, I'll let you search for them
```suggestion
to ensure conntrack is enabled in Neutron server.
```

.. code-block:: yaml

  kolla_neutron_ml2_extension_drivers:
    port_security
    conntrack_helper
    dns_domain_ports

The neutron l3 agent also requires conntrack to be set as an extension in
``kolla/config/neutron/l3_agent.ini``

.. code-block:: ini

  [agent]
  extensions = conntrack_helper

It is also required to load the conntrack kernel module ``nf_nat_tftp``,
``nf_conntrack`` and ``nf_conntrack_tftp`` on network nodes. You can load these

> modules using modprobe or define these in /etc/module-load.

Reviewer: I think the grammar can be improved here a bit. The I think the line lengths are a bit off in my suggestion because I'm doing it in the github comment editor
```suggestion
The conntrack kernel modules ``nf_nat_tftp``, ``nf_conntrack``,
and ``nf_conntrack_tftp`` are also required on network nodes. You
can load these modules using modprobe or define these in /etc/module-load.
```

The Ironic neutron router will also need to be configured to use

> conntrack_helper.

Reviewer: Not completely sure on this one, I'll leave it for you to decide
```suggestion
``conntrack_helper``.
```

.. code-block:: json

  "conntrack_helpers": {
    "protocol": "udp",
    "port": 69,
    "helper": "tftp"
  }

To add the conntrack_helper to the neutron router, you can use the openstack
CLI

.. code-block:: console

  openstack network l3 conntrack helper create \
  --helper tftp \
  --protocol udp \
  --port 69 \
  <ironic_router_uuid>

Baremetal inventory
===================

The baremetal inventory is constructed with three different group types.
The first group is the default baremetal compute group for Kayobe called

> [baremetal-compute] and will contain all baremetal nodes including tenant

Reviewer: I'm now realising that it's a bit redundant of me to keep pointing out these words cat can be codeified. This is the last one I'll do, but please go through the rest of the doc to make sure it all matches up.

```suggestion
``[baremetal-compute]`` and will contain all baremetal nodes including tenant
```

Reviewer: Could you also define what 'tenant' means in this context before you use it? I think I can see what it means from context, but I've not really come across the term before and I'm not sure how common it is.

in Openstack Config.

.. code-block:: yaml

    ironic_resource_class: "example_resource_class"
    ironic_redfish_system_id: "example_system_id"
    ironic_redfish_verify_ca: "{{ inspector_rule_var_redfish_verify_ca }}"

The third group type will be the rack where the node is installed. This is the
group in which the rack specific networking configuration is defined here and
where the BMC address is entered as a host variable for each baremetal node.
Nodes can now be entered directly into the hosts file as part of this group.

.. code-block:: ini

  [rack1]
  hv001 ipmi_address=10.1.28.16
  hv002 ipmi_address=10.1.28.17
  …

This rack group contains the baremetal hosts but will also need to be
associated with the baremetal-compute and baremetal-sr645 groups in order for
those variables to be associated with the rack group.

.. code-block:: ini

  [baremetal-<YOUR_BAREMETAL_HARDWARE_TYPE>:children]
  rack1
  …

  [baremetal-compute:children]
  rack1
  …

Node enrollment
===============

When nodes are defined in the inventory you can begin enrolling them by
invoking the Kayobe commmand

.. code-block:: console

>   (kayobe) $ kayobe baremetal compute register

Reviewer: I'm personally not a fan of these prefixes since it makes it harder to copy commands out
```suggestion
  kayobe baremetal compute register
```

All nodes that were not defined in Ironic previously should’ve been enrolled
following this playbook and should now be in ‘manageable’ state if Ironic was
able to reach the BMC of the node. We will need to inspect the baremetal nodes
to gather information about their hardware to prepare for deployment. Kayobe

> provides an inspection workflow and can be run using:

Reviewer: I think the term `workflow` has the potential to be confused with github workflows, particularly because we're increasingly using kayobe-automation more now
```suggestion
provides an inspection command and can be run using:
```

.. code-block:: console

  (kayobe) $ kayobe baremetal compute inspect

Inspection would require PXE booting the nodes into IPA. If the nodes were able
to PXE boot properly they would now be in ‘manageable’ state again. If an error
developed during PXE booting, the nodes will now be in ‘inspect failed’ state
and issues preventing the node from booting or returning introspection data
will need to be resolved before continuing. If the nodes did inspect properly,

> they can be cleaned and made available to Nova by running the provide workflow.

Reviewer: ```suggestion
they can be cleaned and made available to Nova by running the provide command.
```

.. code-block:: console

  (kayobe) $ kayobe baremetal compute provide

Baremetal hypervisors
=====================

Nodes that will not be dedicated as baremetal tenant nodes can be converted

> into hypervisors as required. StackHPC Kayobe configuration provides a workflow

Reviewer: Again, not sure `workflow` is the best term here because it's already used for other things

to provision baremetal tenants with the purpose of converted these nodes to
hypervisors. To begin the process of converting nodes we will need to define a
child group of the rack which will contain baremetal nodes dedicated to compute
hosts.

.. code-block:: ini

  [rack1]
  hv001 ipmi_address=10.1.28.16
  hv002 ipmi_address=10.1.28.17
  …

  [rack1-compute]
  hv003 ipmi_address=10.1.28.18
  hv004 ipmi_address=10.1.28.19
  …

  [rack1:children]
  rack1-compute

  [compute:children]
  rack1-compute

The rack1-compute group as shown above is also associated with the Kayobe

> compute group in order for Kayobe to run the compute Kolla workflows on these

Reviewer: Again "workflow" can be confusing

nodes during service deployment.

> You will also need to setup the Kayobe network configuration for the rack1

Reviewer: https://promova.com/confusing-words/setup-set-up
```suggestion
You will also need to set up the Kayobe network configuration for the rack1
```

this should consist of the correct CIDR for the rack being deployed.
The configuration should resemble below in networks.yml:

.. code-block:: yaml

  physical_rack1_admin_oc_net_cidr: “172.16.208.128/27”
  physical_rack1_admin_oc_net_gateway: “172.16.208.129”
  physical_rack1_admin_net_defroute: true

You will also need to configure a neutron network for racks to deploy instances
on, we can configure this in openstack-config as before. We will need to define
this network and associate a subnet for it for each rack we want to enroll in
Ironic.

.. code-block:: yaml

  openstack_network_rack:
    name: "rack-net"
    project: "admin"
    provider_network_type: "vlan"
    provider_physical_network: "provider"
    provider_segmentation_id: 450
    shared: false
    external: false
    subnets:
    - "{{ openstack_subnet_rack1 }}"

  openstack_subnet_rack1:
    name: "rack1-subnet"
    project: "admin"
    cidr: "172.16.208.128/27"
    enable_dhcp: false
    gateway_ip: "172.16.208.129"
    allocation_pool_start: "172.16.208.130"
    allocation_pool_end: "172.16.208.130"

The subnet configuration largely resembles the Kayobe network configuration,

> however we do not need to define an allocation pool or enable dhcp as we will

Reviewer: ```suggestion
however we do not need to define an allocation pool or enable DHCP as we will
```

be associating neutron ports with our hypervisor instances per IP address to
ensure they match up properly.

Now we should ensure the network interfaces are properly configured for the
rack1-compute group, the interfaces should include the kayobe admin network
for rack1 and the kayobe internal API network and be defined in the group_vars.

.. code-block:: yaml

 network_interfaces:
  - "internal_net"
  - "physical_rack1_admin_oc_net"

  admin_oc_net_name: "physical_rack1_admin_oc_net"

  physical_rack1_admin_oc_net_bridge_ports:
    - eth0
  physical_rack1_admin_oc_net_interface: br0

  internal_net_interface: "br0.{{ internal_net_vlan }}"

We should also ensure some variables are configured properly for our group,
such as the hypervisor image. These variables can be defined anywhere in

> group_vars, we can place them in the ironic-vars file we used before for

Reviewer: `anywhere in group_vars` is a bit misleading since you need to pick a correct group. I'd probably just remove the word "anywhere"

baremetal node registration.

.. code-block:: yaml

  hypervisor_image: "<image_uuid>"
  key_name: "<key_name>"
  availability_zone: "nova"
  baremetal_flavor: "<ironic_flavor_name>"
  baremetal_network: "rack-net"
  auth:
    auth_url: "{{ lookup('env', 'OS_AUTH_URL') }}"
    username: "{{ lookup('env', 'OS_USERNAME') }}"
    password: "{{ lookup('env', 'OS_PASSWORD') }}"
    project_name: "{{ lookup('env', 'OS_PROJECT_NAME') }}"

With these variables defined we can now begin deploying the baremetal nodes as

> instances, to begin we invoke the deploy-baremetal-instance ansible playbook.

Reviewer: ```suggestion
instances, to begin we invoke the deploy-baremetal-instance Ansible playbook.
```

.. code-block:: console

  kayobe playbook run $KAYOBE_CONFIG_PATH/ansible/deploy-baremetal-instance.yml

This playbook will update the Kayobe network allocations with the the admin
network addresses associated with that rack for each baremetal server, e.g.
in the case of rack 1 this will appear in network-allocations.yml as

.. code-block:: yaml

  physical_rack1_admin_oc_net_ips:
    hv003: 172.16.208.133
    hv004: 172.16.208.134

Once the network allocations have been updated, the playbook will then create a
Neutron port configured with the address of the baremetal node admin network.
The baremetal hypervisors will then be imaged and deployed associated with that
Neutron port. You should ensure that all nodes are correctly associated with
the right baremetal instance, you can do this by running a baremetal node show

> on any given hypervisor node and comparing the server uuid to the metadata on

Reviewer: ```suggestion
on any given hypervisor node and comparing the server UUID to the metadata on
```

the Nova instance.

Once the nodes are deployed, we can use Kayobe to configure them as compute
hosts, running kayobe overcloud host configure on these nodes will ensure that

> all networking, package and various other host configurations are setup

Reviewer: I think the second half of the sentence here is a bit redundant. If you want to explain a bit more about host configuration, I'd just link to the upstream kayobe docs
```suggestion
Once the nodes are deployed, we can use Kayobe to configure them as compute
hosts.
```

.. code-block:: console

  kayobe overcloud host configure --limit baremetal-<YOUR_BAREMETAL_HARDWARE_TYPE>

Following host configuration we can begin deploying OpenStack services to the
baremetal hypervisors by invoking kayobe overcloud service deploy. Nova

> services will be deployed to the baremetal hosts.

Reviewer: ```suggestion
Following host configuration we can begin deploying OpenStack services to the
baremetal hypervisors by invoking ``kayobe overcloud service deploy``.
```

.. code-block:: console

>   kayobe overcloud service deploy --kolla-limit baremetal-<YOUR_BAREMETAL_HARDWARE_TYPE>

Reviewer: This won't add the nodes to the centralised monitoring because the controllers are skipped

Un-enrolling hypervisors
========================

> To convert baremetal hypervisors into regular baremetal compute instances you
> will need to drain the hypervisor of all running compute instances, you should
> first invoke the nova-compute-disable playbook to ensure all Nova services on
> the baremetal node are disabled and compute instances will not be allocated to
> this node.

Reviewer: Very long sentence, needs breaking up a bit
```suggestion
To convert baremetal hypervisors into regular baremetal compute instances you
will need to drain the hypervisor of all running compute instances. 
First invoke the ``nova-compute-disable.yml`` playbook to ensure all Nova services on
the baremetal node are disabled and compute instances will not be allocated to
this node.
```

.. code-block:: console

  (kayobe) $ kayobe playbook run $KAYOBE_CONFIG_PATH/ansible/nova-compute-disable.yml

Now the Nova services are disabled you should also ensure any existing compute
instances are moved elsewhere by invoking the nova-compute-drain playbook

.. code-block:: console

  (kayobe) $ kayobe playbook run $KAYOBE_CONFIG_PATH/ansible/nova-compute-drain.yml

Now the node has no instances allocated to it you can delete the instance using

> the OpenStack CLI and the node will be moved back to ``available`` state.

Reviewer: Maybe just clarify here that it's the baremetal instance that you're deleting, not the VMs that you've migrated away
